File: languages/runtime/cpp/backend_checker.cpp
#ifndef BACKEND_CHECKER_CPP
#define BACKEND_CHECKER_CPP

#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <chrono>
#include <functional>
#include "json.hpp"

using json = nlohmann::json;

namespace BackendChecker {
    struct FailedTestcase {
        int id;
        json input;
        json expected;
        json got;
        std::string stdOut;

        json toJson() const {
            return {
                {"id", id},
                {"input", input},
                {"expected", expected},
                {"got", got},
                {"stdOut", stdOut}
            };
        }
    };

    struct RunResult {
        bool isPassed = true;
        std::vector<FailedTestcase> failedTestcases;
        double memoryUsage = 0.0;
        double executionTime = 0.0;
        std::string stdOut;
        std::string stdErr;

        json toJson() const {
            json failedArray = json::array();
            for (const auto& tc : failedTestcases) {
                failedArray.push_back(tc.toJson());
            }

            return {
                {"isPassed", isPassed},
                {"failedTestcases", failedArray},
                {"memoryUsage", memoryUsage},
                {"executionTime", executionTime},
                {"stdOut", stdOut},
                {"stdErr", stdErr}
            };
        }
    };

    template<typename T>
    struct StdoutCapture {
        T result;
        std::string stdout;
    };

    template<typename Func>
    auto captureStdout(Func fn) -> StdoutCapture<decltype(fn())> {
        std::stringstream buffer;
        std::streambuf* old = std::cout.rdbuf(buffer.rdbuf());
        
        auto result = fn();
        
        std::cout.rdbuf(old);
        
        return {result, buffer.str()};
    }

    void writeResults(const RunResult& results, const std::string& filename) {
        std::ofstream file(filename);
        if (!file.is_open()) {
            std::cerr << "Failed to write results file." << std::endl;
            exit(1);
        }
        file << results.toJson().dump();
        file.close();
    }

    template<typename T, typename Func>
    void runTests(const std::vector<T>& testcases, Func runChecker) {
        auto startTime = std::chrono::high_resolution_clock::now();

        RunResult results = runChecker(testcases);

        auto endTime = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(endTime - startTime);
        results.executionTime = static_cast<double>(duration.count());

        writeResults(results, "results.json");
    }
}

#endif

File: languages/runtime/java/BackendChecker.java
import com.google.gson.Gson;
import java.io.ByteArrayOutputStream;
import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintStream;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;

public class BackendChecker {
  public static class FailedTestcase {
    public int id;
    public Object input;
    public Object expected;
    public Object got;
    public String stdOut;
  }

  public static class RunResult {
    public boolean isPassed = true;
    public List<FailedTestcase> failedTestcases = new ArrayList<>();
    public double memoryUsage;
    public double executionTime;
    public String stdOut;
    public String stdErr;
  }

  public static class StdoutCapture {
    public Object result;
    public String stdout;
    
    public StdoutCapture(Object result, String stdout) {
      this.result = result;
      this.stdout = stdout;
    }
  }

  public static <T> StdoutCapture captureStdout(java.util.function.Supplier<T> fn) {
    ByteArrayOutputStream baos = new ByteArrayOutputStream();
    PrintStream oldOut = System.out;
    System.setOut(new PrintStream(baos));
    
    T result = fn.get();
    
    System.setOut(oldOut);
    String stdout = baos.toString();
    
    return new StdoutCapture(result, stdout);
  }

  public static void writeResults(RunResult results, String filename) {
    try (FileWriter file = new FileWriter(filename)) {
      Gson gson = new Gson();
      file.write(gson.toJson(results));
    } catch (IOException e) {
      System.err.println("Failed to write results file.");
      System.exit(1);
    }
  }

  public static <T> void runTests(List<T> testcases, Function<List<T>, RunResult> runChecker) {
    long startTime = System.currentTimeMillis();

    RunResult results = runChecker.apply(testcases);

    long endTime = System.currentTimeMillis();
    results.executionTime = (double) (endTime - startTime);

    writeResults(results, "results.json");
  }

  public static void main(String[] args) {
    List<Driver.Testcase> testcases = Driver.loadTestcases("testcases.json");
    runTests(testcases, MainChecker::runChecker);
  }
}
